src/lib/support/BytesCircularBuffer.cpp

```diff
@@ -0,0 +1,158 @@
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <support/BytesCircularBuffer.h>
+
+#include <limits>
+#include <nlassert.h>
+#include <string.h>
+
+#include <support/CodeUtils.h>
+
+namespace chip {
+
+size_t BytesCircularBuffer::Advance(size_t dataLocation, size_t amount) const
+{
+    dataLocation += amount;
+    if (dataLocation >= mCapacity)
+        dataLocation -= mCapacity;
+    return dataLocation;
+}
+
+void BytesCircularBuffer::Read(uint8_t * dest, size_t length, size_t offset) const
+{
+    VerifyOrDie(StorageUsed() >= offset + length);
+
+    size_t start = Advance(mDataStart, offset);
+    if (mCapacity - start >= length)
+    {
+        ::memcpy(dest, mStorage + start, length);
+    }
+    else
+    {
+        size_t firstPiece  = mCapacity - start;
+        size_t secondPiece = length - firstPiece;
+        ::memcpy(dest, mStorage + start, firstPiece);
+        ::memcpy(dest + firstPiece, mStorage, secondPiece);
+    }
+}
+
+void BytesCircularBuffer::Write(const uint8_t * source, size_t length)
+{
+    // Always reserve 1 byte to prevent mDataStart == mDataEnd because then it would be
+    // ambiguous whether we have 0 bytes or mCapacity bytes stored.
+    VerifyOrDie(mCapacity - StorageUsed() - 1 >= length);
+
+    if (mCapacity - mDataEnd >= length)
+    {
+        ::memcpy(mStorage + mDataEnd, source, length);
+        mDataEnd = Advance(mDataEnd, length); // Use advance in case that mDataEnd wrap to 0
+    }
+    else
+    {
+        size_t firstPiece  = mCapacity - mDataEnd;
+        size_t secondPiece = length - firstPiece;
+        ::memcpy(mStorage + mDataEnd, source, firstPiece);
+        ::memcpy(mStorage, source + firstPiece, secondPiece);
+        mDataEnd = secondPiece;
+    }
+}
+
+void BytesCircularBuffer::Drop(size_t length)
+{
+    VerifyOrDie(StorageUsed() >= length);
+    mDataStart = Advance(mDataStart, length);
+}
+
+size_t BytesCircularBuffer::StorageUsed() const
+{
+    if (mDataStart <= mDataEnd)
+    {
+        return mDataEnd - mDataStart;
+    }
+    else
+    {
+        return mCapacity + mDataEnd - mDataStart;
+    }
+}
+
+CHIP_ERROR BytesCircularBuffer::Push(const ByteSpan & payload)
+{
+    size_t maxStorageAvailable = mCapacity - (sizeof(SizeType) + 1);
+    size_t length              = payload.size();
+    if (length > maxStorageAvailable)
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    if (length > std::numeric_limits<SizeType>::max())
+        return CHIP_ERROR_INVALID_ARGUMENT;
+
+    // Free up space until there is enough space.
+    while (length > maxStorageAvailable - StorageUsed())
```

Comment:
OK.  So let's say mCapacity is 10 and StorageUsed() is 8.  maxStorageAvailable will be 7, this line will test `length > size_t(-1)` which will of course test false and we won't Pop anything at all, which is not the right behavior.  This should have been covered by the tests I asked for last time...